### ISSUE TITLE
Merge Pred and Expr

### ISSUE THREAD

**Issue by spinda** [ucsd-progsys/liquid-fixpoint#146]
We've talked about this before, so I'm submitting this ticket to make sure it gets done eventually. `Pred` and `Expr` should be merged (into `Logic`?) to lift some weird arbitrary restrictions (eg. you can't have a `Pred` in an application).

**Comment by nikivazou** [comment]
Sure! I am favoring merging.

**Comment by ranjitjhala** [comment]
Is there any reason NOT to merge?

**Comment by ranjitjhala** [comment]
Ie if all tests pass I am happy to merge

**Comment by nikivazou** [comment]
The only issue is that the interface will break.

**Comment by ranjitjhala** [comment]
We can bump up the version number etc to take care of that (I hope you all followed the other thread on that topic?

**Comment by spinda** [comment]
To start off, I'll copy/paste some code from an email I sent about this a while back:

```haskell
data Logic      = LBot
                | LProp  !Prop
                | LConst !Constant
                | LVar   !LocSymbol

                | LBin   !Bop
                | LAtom  !Brel
                | LConn  !Connective

                | LNot   !Logic
                | LNeg   !Logic
                | LIte   !Logic !Logic !Logic

                | LCst   !Logic !Sort
                | LAll   ![(Symbol, Sort)] !Logic

                | LApp   !Logic !Logic
                deriving (Eq, Ord, Show, Data, Typeable, Generic)

data Prop       = PTrue | PFalse | PTop
                deriving (Eq, Ord, Show, Data, Typeable, Generic)
data Constant   = I !Integer | R !Double | L !Text
                deriving (Eq, Ord, Show, Data, Typeable, Generic)
data Bop        = Plus | Minus | Times | Div | Mod
                deriving (Eq, Ord, Show, Data, Typeable, Generic)
data Brel       = Eq | Ne | Gt | Ge | Lt | Le | Ueq | Une
                deriving (Eq, Ord, Show, Data, Typeable, Generic)
data Connective = And | Or | Iff | Imp
                deriving (Eq, Ord, Show, Data, Typeable, Generic)
```

Some things are missing from here, and it's probably not exactly how the final representation needs to be, so&mdash;feedback/changes?

**Comment by gridaphobe** [comment]
Super nitpicky: most of these datatypes have no sensible definition for `Ord`.

**Comment by spinda** [comment]
True&mdash;I'm not sure why I put that in there...

**Comment by christetreault** [comment]
Some feedback from a mere mortal; Some of these types don't have names that are super obvious to the untrained eye. `Prop` and `Constant`

**Comment by gridaphobe** [comment]
More nitpicky comments:

1. I agree with @christetreault (even though I'm also guilty of using `Bop`), full words are better.
2. I really dislike this habit we have of using a single char prefix for certain datacons, e.g. `LProp` etc. We have a module system, let's use it!
3. Why does `Top :: Prop` but `Bot :: Logic`? Strange asymmetry there..
4. Neither fixpoint nor LH use `Top` or `Bot` **anywhere**, can we kill them? Shouldn't `Top` be `True` and `Bot` `False` anyway?
5. `Logic` is the wrong name for the type. `true` is not a logic, it's an expression (or a term) in a logic.
6. Since the expressions are now untyped, does it make sense to separate logical connectives, arithmetic operators, and relational operators? They're all just binary term constructors now.

**Comment by spinda** [comment]
Revised version, taking into account @christetreault's and @gridaphobe's comments and adding in some new ideas.

```haskell
{-# LANGUAGE OverloadedStrings #-}

module Language.Fixpoint.Types (
    Expr(..)
  , Literal(..)
  , Operator(..)
  , KVar(..)
  , Variable
  , mkVar
  , mkVarAnn
  , varSymbol
  , varAnnotation
  , annotateVar
  ) where

import Data.Dynamic
import Data.Typeable (Typeable)

import Language.Fixpoint.Names

-- Expression AST --------------------------------------------------------------

data Expr       = Literal  !Literal
                | Operator !Operator
                | Variable !(Located Variable)

                | IfThenElse !Expr !Expr !Expr

                | Cast !Expr !Sort
                | ForAll ![(Symbol, Sort)] !Expr

                | KVar !KVar !Subst

                | Apply ![Expr]
                deriving (Eq, Data, Typeable, Generic)

data Literal    = Bool     !Bool
                | Integer  !Integer
                | Real     !Double
                | String   !Symbol
                | Uninterp !Symbol !Sort
                deriving (Eq, Data, Typeable, Generic)

data Operator   = Not | Neg
                | Plus | Minus | Times | Div
                | Mod | Eq | Ne | Gt | Ge | Lt | Le | Ueq | Une
                | And | Or | Iff | Imp
                deriving (Eq, Data, Typeable, Generic)

-- KVars -----------------------------------------------------------------------

newtype KVar    = KV { kv :: Symbol }
                deriving (Eq, Data, Typeable, Generic, IsString)

-- Variables -------------------------------------------------------------------

data Variable   = V { var_symbol     :: !Symbol
                    , var_annotation :: !(Maybe Dynamic)
                    }
                deriving (Eq, Data, Typeable, Generic)

instance IsString Variable where
  fromString = mkVar

mkVar :: Symbolic s => s -> Variable
mkVar = (`V` Nothing) . symbol

mkVarAnn :: (Symbolic s, Typeable a) => s -> a -> Variable
mkVarAnn sym ann = V (symbol s) (toDyn ann)

varSymbol :: Variable -> Symbol
varSymbol = var_symbol

varAnnotation :: Typeable a => Variable -> Maybe a
varAnnotation = fromDynamic . var_annotation

annotateVar :: Typeable a => Variable -> a -> Variable
annotateVar var ann = var { var_annotation = toDyn ann }
```

In this version, variables in Fixpoint can have arbitrary annotations attached alongside the symbols in the AST (see [`Data.Dynamic`](https://hackage.haskell.org/package/base-4.8.1.0/docs/Data-Dynamic.html)). This allows for carrying around language-specific information (eg. the `Var`s of measures in LiquidHaskell) while keeping Fixpoint mostly language-agnostic.

(Today @nikivazou, @gridaphobe, @bmcfluff, and I talked about parameterizing the `Expr` and `Pred` structs to accomplish something similar, but in implementing that, I found it makes `Subable` and `Subst` a bit weird.)

**Comment by christetreault** [comment]
Seems reasonable to me. Only issue that remains for me is `KVar`, but I understand that this is what this thing has been called since time immemorial, so I'm not going to make a stink over it since changing this will require refactoring a lot of code.

Also; why snake_case in the fields of `Variable`?

BTW, what _does_ the `K` in `KVar` stand for?

**Comment by gridaphobe** [comment]
The `K` comes from the original liquid types paper (PLDI 2008), which uses a fancy lowercase `k` to denote liquid type variables (though it's not in the syntax figure, tsk tsk). As to why `k` instead of eg `p` for predicate, I don't know, there may be precedent in the literature.

**Comment by christetreault** [comment]
Ahh, well if it's `kvar` in the original paper, then far be it from me to question it! :)

**Comment by spinda** [comment]
@christetreault Maybe this is more common in LiquidHaskell than Fixpoint, but in LH many data types use GHC's `prefix_camelCase` convention for field names. Can change if it doesn't fit with Fixpoint's code.

**Comment by ranjitjhala** [comment]
FWIW I greatlyPreferThisStyle to_this_one (as does hlint) so if we're
making changes, and its easy, go for the former...


On Thursday, October 1, 2015, Michael Smith <notifications@github.com>
wrote:

> @christetreault <https://github.com/christetreault> Maybe this is more
> common in LiquidHaskell than Fixpoint, but in LH many data types use GHC's
> prefix_camelCase convention for field names. Can change if it doesn't fit
> with Fixpoint's code.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ucsd-progsys/liquid-fixpoint/issues/146#issuecomment-144791617>
> .
>


-- 
Ranjit.

**Comment by christetreault** [comment]
"many data types" is certainly a sticking point for me. It should be either "all" or "none". Though once again fixing this is outside the scope of what we're talking about.

That said, I haven't noticed snake_case before, so I'm inclined to believe that more types don't use it, so maybe we shouldn't add more that do?

Haskell is pretty strongly in the camelCase convention, so I'd personally perfer we used it.

**Comment by spinda** [comment]
Just to clarify, the naming style I'm referring to isn't `all_snake_case`, it's `prefix_camelCaseName`, where the `prefix_` keeps the fields from clashing with those of other types. Either style is fine with me&mdash;if `camelCaseOnly` is better here then we'll go with that.

Any more feedback on the structure of the AST? Any pieces I'm missing? (eg. should we add more location info?)

**Comment by bmcfluff** [comment]
Resolved by #178
